containerd integration: handle multi-platform images

laurazard opened the pull request:
Signed-off-by: Laura Brehm <laurabrehm@hey.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Fix listing images with containerd store

Upstreams:
- https://github.com/rumpl/moby/pull/113
- https://github.com/rumpl/moby/pull/126

**- How I did it**

Multiple entries are returned for each platform of an image (more discussion here: https://github.com/rumpl/moby/pull/113) – which isn't fantastic, but is better than the current state where listing images is broken.

**- How to verify it**

- `docker pull --platform=linux/amd64 alpine:latest`
- `docker images` (verify it doesn't break)
- `docker pull --platform=[another platform] alpine:latest`
- `docker images` (verify another entry is returned)

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

<img width="491" alt="image" src="https://user-images.githubusercontent.com/70572044/218497498-ae606a32-2521-4855-8be5-bed94b3f680a.png">

laurazard committed:
`docker image ls`: handle multi-platform images

Multiple entries are returned for each platform of an image

Signed-off-by: Laura Brehm <laurabrehm@hey.com>

tianon commented:
This is likely good (for now?), but we probably also want to come to a decision on https://github.com/moby/moby/issues/44582 before we get _too_ deep. :sweat_smile:

(I think this is on me and other maintainers -- we need to participate more over there and I'll go leave some of my thoughts we'd talked about in a previous maintainer meeting over there :see_no_evil:)

To be clear, I think _something_ beats nothing here, and do _not_ personally believe that should be a blocker for this PR (because I don't think any implementation here is going to make it harder one way or another to implement something else). :+1:

laurazard commented:
@tianon I agree! I made this because I was running into not being able to list my images while refactoring for the other PRs to solve the platform issues associated with https://github.com/moby/moby/pull/44958 and https://github.com/moby/moby/pull/44934. 

This PR just makes things a little smoother for now, and we can come to a better solution from discussion in https://github.com/moby/moby/issues/44582.

neersighted reviewed (changes requested):
We discussed this extensively in the context of https://github.com/moby/moby/pull/44840 -- there were a couple takeaways:

These PRs both introduce `isDanglingImage` and `<none>:<none>` strings -- so we'd like to see the common parts factored out into a third PR that these two can then be re-stacked on.

In particular, centralizing the `<none>` string constants in the presentation layer and using a sentinel value (possibly the empty string) to represent untagged images in the plumbing layer makes a lot of sense. We should quarantine those values to the API, so that we can possibly deprecate them in favor of client-side presentation in a future API version.

Likewise, `isDanglingImage` should be factored into a common place for both these PRs (and then consumed by restacking on top of the common PR), as otherwise when both are merged the symbols will collide as they're in the same package.

laurazard commented:
As discussed in https://github.com/moby/moby/pull/44840#issuecomment-1440327282 (and since it's been merged in the meantime), I rebased this PR on top of `master` and addressed @neersighted's comments. PTAL :)

neersighted reviewed (approved):
LGTM!

thaJeztah commented:
First PR in; congrats!
